### ISSUE TITLE
sea_surface marking + costmap dynamics: accumulate per-pixel softmax log-odds (drop the argmax gate), graded cost ramp, bounded recovery

### ISSUE THREAD

**Issue by rolker** [rolker/unh_marine_perception#22]
## Summary

The sea-surface costmap collapses a **graded per-pixel softmax** into a hard
**argmax classification**, then votes a flat `±` log-odds. Two failure modes follow:

1. **Weak-but-persistent obstacles never register.** A pixel where `P(obstacle)`
   is elevated but not the argmax class (e.g. a dim mooring ball under dark skies)
   is scored as **confident water (a `miss`)**, *actively eroding* accumulated
   obstacle evidence. The operator saw a red buoy mark in the segmentation image
   that never appeared on the costmap (deployment
   rolker/unh_echoboats_project11#186, 2026-05-28). Because the **Collision Monitor
   reflex feed shares the same gate**, the target is invisible to the live safety
   layer too — not just the planner.

2. **Free-space recovery is over-damped.** A long-observed water cell saturates to
   the negative clamp (−5.0); a newly-appeared obstacle there must climb +6.0 at
   +0.85/hit ≈ **8 frames (~1.6 s at 5 fps)** to mark, vs **2 frames** for a
   never-observed cell — a 4× penalty from the symmetric clamp.

## Verified mechanism (source)

**The segmentation output is calibrated, not hard-labeled.**
`SegmentorCamera::segmentationCallback`
(`sea_surface_segmentation/src/sea_surface_segmentation.cpp:124–131`) applies a
per-pixel **softmax** over 3 class logits and scales to 255:

```cpp
sum = Σ_j exp(logit_j);   channel_j = 255 * exp(logit_j) / sum;
```

So **R = 255·P(obstacle), G = 255·P(water), B = 255·P(sky)** and
**R + G + B = 255**. The channel magnitude *is* the network's per-class confidence.

**The costmap throws that away:**
- **Gate** — `is_obstacle_pixel` is a strict argmax `R > G && R > B`
  (`segments_projection.hpp:44`); discards the probability it's built on.
- **Vote** — a passing cell produces a flat `hit(+0.85)`, a failing cell a flat
  `miss(−0.40)` (`sea_surface_layer.cpp:408–414`). Magnitude unused.
- The inverse path's `OccupancyObservation` is `{x, y, bool obstacle}`
  (`segments_projection.hpp:~245`) — the intensity isn't even carried into the
  costmap path (the forward `ProjectedPoint` carries it, but only the reflex
  pointcloud consumes that).
- **Buffer** (`occupancy_buffer.hpp`): `hit=+0.85, miss=−0.40, clamp=±5.0,
  lethal_threshold=1.0, decay_half_life=30 s`.
- **Relay** (`sea_surface_relay_layer.cpp`): forwards only cells `≥100` (lethal)
  to the global costmap.

Net: a near-tie softmax (e.g. `P(obs)=0.34, P(water)=0.35`) becomes a confident
−0.40 miss — a 1.25 log-odds swing the wrong way across a one-count boundary,
with no graded middle.

## Proposed design (converged)

**A. Replace the argmax gate + flat vote with the real per-pixel log-odds.**
Because the mask is already a softmax, `P(obs)=R/255` and
`1−P(obs)=(255−R)/255`, so the textbook occupancy increment is:

```
Δ = clamp( log( R / (255 − R) ) )
```

Feed that into the existing log-odds buffer per observation instead of a flat
`±const`. A marginal-red pixel nudges slightly positive; a confident one banks
fast; a water-dominant pixel votes proportionally negative — never the flat
cliff. This **subsumes the old "gate + graded vote"** into one principled step;
clamp / decay / threshold all stay. Requires threading the value into
`OccupancyObservation` (a confidence/log-odds field).
- Guard the tails (`R∈{0,255}` → ±∞): clamp per-observation Δ and/or evaluate
  `log((R+ε)/(255−R+ε))`.

**B. Bound recovery latency — asymmetric clamp.** Floor free-space at a shallow
negative (e.g. −2.0) while letting obstacle evidence bank to +5.0. Recovery from
"cleared" to lethal drops from ~8 hits to ~4 (~0.8 s) while flicker rejection and
durable static persistence are preserved.

**C. Graded cost ramp output (resolves the dead "unknown" zone).** Map
accumulated log-odds to a monotonic cost instead of binary lethal-or-nothing:

```
a ≤ clear_thresh      → 0    (FREE)
clear < a < lethal    → 1…252 ramp (soft "prefer to avoid", not prohibited)
a ≥ lethal_thresh     → 254  (LETHAL, hard no-go)
never observed (NaN)  → don't write (255 / no opinion)
```

"Unknown" then means *only* never-observed; uncertain-with-some-evidence gets a
real cost the planner bends around (`SmacPlannerHybrid` `cost_penalty`; MPPI
`CostCritic` honor it — strength is a joint tune with our ramp). Soft costs stay
≤252 to avoid being read as `INSCRIBED` (253). **The relay must forward the full
gradient** (not just `≥100`) or the global planner is blind to soft costs and the
routing benefit is lost. Inflation only wraps the `LETHAL` tops (soft cells get a
planner bias but no footprint reservation — intended).

**Reflex-path consistency.** The softmax signal is the common currency: the
Collision Monitor pointcloud (instantaneous, binary) should threshold
`P(obs) > τ` (replacing strict argmax) so faint obstacles become visible to safety
too. The reflex monitor remains the **independent binary safety floor**; the
graded costmap is **soft planning preference**. Confident → lethal in both; faint
→ soft cost only (deliberately below `τ` to avoid emergency-stopping on noise —
itself a safety-tuning call).

## Decided against (record)

**Two-timescale / transient-obstacle channel rebuild — not pursued.** The
Collision Monitor reflex path is live and owns dynamic-obstacle safety, so the
costmap stays a deliberative *static-obstacle* map with the 30 s decay. Revisit
only if the reflex path is removed.

## Optional follow-on (not blocking)

Publish the obstacle-class **log-odds directly from the node** (`layer_data`
already holds the pre-softmax logits) to avoid 8-bit quantization loss in the
confident tail. Reconstructing `log(R/(255−R))` from the rgb8 mask is fine
meanwhile since we clamp.

## Validate / tune (cheap, against #186 bag)

Mechanism is confirmed from source (softmax → graded). Remaining empirical step is
**threshold tuning**: sample R at the false-negative pixels (operator saw red, no
mark) and at open-water in
`~/data/logs/bizzy_images/bag_2026-05-28T13.25.54_ffmpeg_seg/` to set
`clear_thresh` / `lethal_thresh` / `τ` against the real R distribution.

## Scope

- **In scope:** per-observation evidence (`log(R/(255−R))`), buffer clamp
  asymmetry, graded cost-ramp output + relay gradient, reflex-gate consistency.
- **Out of scope:** dark-sky cases where the model emits no obstacle probability at
  all (camera / model / exposure concern).
- **Orthogonal (separate):** close-range per-cell dilution — many water pixels
  diluting few buoy pixels in one cell (#186 §10).

---
**Authored-By**: `Claude Code Agent`
**Model**: `Claude Opus 4.8 (1M context)`
